File: chap5/sudoku.cpp
// i was attempting to use python to implement a sudoku solver
// and now, after understanding more about backtracking,
// i decided to reimplement the sudoku solver in C++
//
// C++11


#include <bits/stdc++.h>
using namespace std;

bool valid(vector<vector<int>> sudoku) {
    // check row
    for (auto i : sudoku) {
        vector<int> checkrow;
        for (auto j : i) {
            if (j != 0) {
                checkrow.push_back(j); 
            }
        }
        unordered_set<int> rowset(checkrow.begin(), checkrow.end());
        if (rowset.size() != checkrow.size()) {
            return false; 
        }
    }
    // check column
    for (int i = 0; i < 9; i++) {
        vector<int> checkcol;
        for (auto j: sudoku) {
            if (j[i] != 0) {
                checkcol.push_back(j[i]);
            }
        }
        unordered_set<int> colset(checkcol.begin(), checkcol.end());
        if (checkcol.size() != colset.size()) {
            return false;
        }
    }
    // check grid
    for (int i = 0; i <= 6; i += 3) {
        for (int j = 0; j <= 6; j += 3) {
            vector<int> box;
            for (int x = i; x < i + 3; x++) {
                for (int y = j; y < j + 3; y++) {
                    if (sudoku[x][y] != 0) {
                        box.push_back(sudoku[x][y]);
                    }
                }
            }
            unordered_set<int> boxset(box.begin(), box.end());
            if (box.size() != boxset.size()) {
                return false;
            }
        }
    }
    
    return true;
}

tuple<int, int> nav(int row, int col) {
    if (col >= 8) {
        return make_tuple(row + 1, 0);
    } else {
        return make_tuple(row, col + 1);
    }
}

bool solve(vector<vector<int>>& sudoku, int row, int col)
{
    if (row == 9) {
        return true;
    }
    else {
        if (sudoku[row][col] != 0) {
            int r, c;
            tuple<int, int> t = nav(row, col);
            tie(r, c) = t;
            if (solve(sudoku, r, c)) {
                return true;
            }
        }
        for (int i = 1; i <= 9; i++) {
            sudoku[row][col] = i;
            if (valid(sudoku)) {
                int r, c;
                tuple<int, int> t = nav(row, col);
                tie(r, c) = t;
                if (solve(sudoku, r, c)) {
                    return true;
                }
            }
            sudoku[row][col] = 0;
        }
    }
    return false;
}

void printSudoku(vector<vector<int>>& sudoku) {
    for (auto& row : sudoku) {
        for (int num : row) {
            cout << num << " ";
        }
        cout << endl;
    }
}

int main() {
    vector<vector<int>> sudoku = {
        {0, 0, 5, 8, 0, 9, 4, 7, 0},
        {0, 2, 0, 0, 6, 0, 0, 0, 1},
        {0, 7, 0, 0, 0, 0, 0, 6, 9},
        {4, 0, 0, 0, 0, 0, 1, 0, 0},
        {0, 0, 2, 1, 0, 8, 9, 0, 6},
        {3, 0, 0, 9, 4, 0, 0, 2, 8},
        {2, 3, 0, 7, 5, 0, 6, 9, 4},
        {5, 0, 0, 0, 0, 3, 8, 0, 0},
        {6, 9, 0, 0, 0, 2, 3, 5, 0},
    };

    solve(sudoku, 0, 0);
    printSudoku(sudoku);
    return 0;

}

File: chap5/sudoku.py
# my weird attempt to solve sudoku in python
# again, this is for the purpose of learning backtracking
# which is relevant to the current chapter
#
# also, the reason why I chose python
# is because I did try to use python to solve sudoku
# years ago so it's easier for me to redo it again
# in python since I did remember the general way to do it
# also python is easier to read :)
# sudoku somehow
from pprint import pprint
sudoku = [
    [0, 0, 5, 8, 0, 9, 4, 7, 0],
    [0, 2, 0, 0, 6, 0, 0, 0, 1],
    [0, 7, 0, 0, 0, 0, 0, 6, 9],
    [4, 0, 0, 0, 0, 0, 1, 0, 0],
    [0, 0, 2, 1, 0, 8, 9, 0, 6],
    [3, 0, 0, 9, 4, 0, 0, 2, 8],
    [2, 3, 0, 7, 5, 0, 6, 9, 4],
    [5, 0, 0, 0, 0, 3, 8, 0, 0],
    [6, 9, 0, 0, 0, 2, 3, 5, 0],
]
    
def valid(sudoku):
    # check row
    for i in sudoku:
        checkrow = [e for e in i if e != 0]
        if len(checkrow) != len(set(checkrow)):
            return False
    # check column
    for i in range(0, 9):
        reccol = []
        for j in sudoku:
            reccol.append(j[i])
        checkcol = [e for e in reccol if e != 0] 
        if len(checkcol) != len(set(checkcol)):
            return False
    # check 3x3 box
    for i in [0, 3, 6]:
        for j in [0, 3, 6]:
            box = [sudoku[r][c] for c in range(j, j + 3) for r in range(i, i + 3)]
            checkbox = [e for e in box if e != 0]
            if len(checkbox) != len(set(checkbox)):
                return False
    return True
def nav(row, col):
    if col >= 8:
        return row + 1, 0
    else:
        return row, col + 1

def solve(sudoku, row, col):
    if row == 9:
        return sudoku
    else:
        if sudoku[row][col] != 0:
            nrow, ncol = nav(row, col)
            if solve(sudoku, nrow, ncol):
                return sudoku
        else:
            for i in range(1, 10):
                sudoku[row][col] = i
                if valid(sudoku):
                    nrow, ncol = nav(row, col)
                    if solve(sudoku, nrow, ncol):
                        return sudoku
                sudoku[row][col] = 0

pprint(solve(sudoku, 0, 0))